clang/lib/CIR/Dialect/Transforms/HoistAllocas.cpp

```diff
@@ -49,16 +70,27 @@ static void process(cir::FuncOp func) {
   mlir::Operation *insertPoint = &*entryBlock.begin();
 
   for (auto alloca : allocas) {
-    alloca->moveBefore(insertPoint);
     if (alloca.getConstant()) {
-      // Hoisted alloca may come from the body of a loop, in which case the
-      // stack slot is re-used by multiple objects alive in different iterations
-      // of the loop. In theory, each of these objects are still constant within
-      // their lifetimes, but currently we're not emitting metadata to further
-      // describe this. So for now let's behave conservatively and remove the
-      // const flag on nested allocas when hoisting them.
-      alloca.setConstant(false);
+      if (isOpInLoop(alloca)) {
```

Comment:
Can all this logic be tied to -O1 and above? Seems potentially expensive to happen all the time?

Reply:
Perhaps isolate all this logic into another function

Reply:
Updated.